src/web_app/js/call.js

```diff
@@ -89,19 +131,45 @@ Call.prototype.hangup = function() {
   // When the other client sees BYE it attempts to post offer and candidates to
   // GAE. GAE needs to know that we're disconnected at that point otherwise
   // it will forward messages to this client instead of storing them.
-  var path = this.roomServer_ + '/leave/' + this.params_.roomId +
-      '/' + this.params_.clientId;
-  var xhr = new XMLHttpRequest();
-  xhr.open('POST', path, false);
-  xhr.send();
 
-  // Send bye to other client.
-  this.channel_.send(JSON.stringify({type: 'bye'}));
-  this.channel_.close();
+  // This section of code is executed in both sync and async depending on
+  // where it is called from. When the browser is closed, the requests must
+  // be executed as sync to finish before the browser closes. When called
+  // from pressing the hang up button, the requests are executed async.
+
+  // If you modify the steps used to hang up a call, you must also modify
+  // the clean up queue steps set up in createCleanupQueue_.');
+  var path = this.createLeaveUrl_();
 
-  this.params_.previousRoomId = this.params_.roomId;
-  this.params_.roomId = null;
-  this.params_.clientId = null;
+  var sendLeave = sendUrlRequest('POST', path, async).catch(function(error) {
+    trace('Error sending /leave: ' + error.message);
+  }.bind(this));
+
+  var sendBye = sendLeave.then(function() {
+    // Send bye to the other client.
+    this.channel_.send(JSON.stringify({type: 'bye'}));
+  }.bind(this)).catch(function(error) {
+    trace('Error sending bye: ' + error.message);
+  }.bind(this));
+
+  var closeChannel = sendBye.then(function() {
+    return this.channel_.close(async);
+  }.bind(this)).catch(function(error) {
+    trace('Error closing signaling channel: ' + error.message);
+  }.bind(this));
+
+  var setParams = closeChannel.then(function() {
+    this.params_.previousRoomId = this.params_.roomId;
+    this.params_.roomId = null;
+    this.params_.clientId = null;
+  }.bind(this));
+
```

Comment:
Could you clarify what you mean? The sync mode only runs when the tab/window is closed.